job-info: Return ENODATA when reaching the end of eventlogs

chu11 opened the pull request:
Per discussion #2338.  This is built on top of #2366.

- It is easy to "reach the end" of
  - the main job eventlog (event "clean")
  - a guest eventlog in the guest namespace
  - a guest eventlog copied into the primary KVS namespace

- It is hard to know how to "reach the end" of a hypothetical namespace in the primary KVS that we don't know about.  So I did not support that case.  In that situation, the wait-event watches until it receives a cancellation notice.

- I don't return any events after "clean" in the main job eventlog.  So if the user were to do something stupid like, "flux kvs eventlog append" on it, that event won't show up via an eventlog watch.

- As a side effect of this work #2356 and #2361 are basically closed too.

- Once finishing this work, removed the now unncessary calls to "eventlog_watch_cancel" in `cmd/flux-job`.

- I realized there is a race condition (although never seen) where an entry placed into a guest eventlog will not be seen by job-info.  The race is:

  - user write to eventlog, kvs completes write, sends setroot event
  - user deletes namespace, kvs sends namespace-remove event kvs-watch
  - kvs-watch receives setroot event, does lookupat on key
  - kvs-watch receives namespace-remove event BEFORE lookupat returns, sends ENOTSUP to watchers
  - response from lookupat arrives after ENOTSUP sent

  - I elected to solve this by having watched eventlogs in the guest namespace ALWAYS fallthrough to the main KVS namespace afterwards.  Some offsets checks ensure that all data in the guest eventlog has been read, and if it hasn't, send it over.

chu11 commented:
rebased on master and re-organized so that this can go in without #2366.

garlick commented:
> It is hard to know how to "reach the end" of a hypothetical namespace in the primary KVS that we don't know about. So I did not support that case. In that situation, the wait-event watches until it receives a cancellation notice.

To clarify:  the rule from RFC 21 is that the job directory is no longer written to after the job reaches INACTIVE state, so in theory you could return ENODATA on any eventlog under the job directory once INACTIVE has been reached.

chu11 commented:
> To clarify: the rule from RFC 21 is that the job directory is no longer written to after the job reaches INACTIVE state, so in theory you could return ENODATA on any eventlog under the job directory once INACTIVE has been reached.

Understood.  "hard to know" wasn't probably the right word, perhaps I should have said, "it's a lot of work to know", monitor job state transitions, but that's only if job is active, so gotta check main eventlog first (I think?).

garlick commented:
Pondering if there's anything job manager can do to make life easier here...

Maybe that doesn't have to hold this up though.

garlick commented:
I Have to head home but I'll plan to review some more tonight/tomorrow AM.

chu11 commented:
checked diff coverage, 79% seems good.  Missed "impossible" errors or very racy bits.

re-pushed, just noticed a typo in a comment.

garlick commented on src/modules/job-info/guest_watch.c:
```diff
@@ -626,15 +626,18 @@ static void guest_namespace_watch_continuation (flux_future_t *f, void *arg)
-             * events in the guest eventlog, assume job was moved into
-             * the main namespace before we began watching in the
-             * guest namespace.
+            /* Guest namespace has been removed.  If we have read
```
Commit a3967b7 is mostly overwritten by later commits (squash?)

garlick replied:
The overwritten comment talks about handling the case of an empty eventlog.  I don't think that can happen.  Eventlogs are created when the first event is appended (perhaps that needs calling out in RFC 18?)

chu11 replied:
Yeah, it is overwritten within this commit series.  I decided to leave the commit series b/c of how some individual issues were fixed along the way.  Could squash.

I assumed there could be cases of `mkdir` without an append.  If that's not possible, could close that issue and squash together.  Let me see about how to re-order squash the commits.

garlick replied:
> I assumed there could be cases of mkdir without an append.

Did you mean an empty put?

chu11 replied:
> Did you mean an empty put?

What I meant to say was "mkdir and nothing written afterwards".  But yeah, an empty put.

chu11 commented:
> Pondering if there's anything job manager can do to make life easier here...

could job state transitions be put into the eventlog?  or perhaps another eventlog with a known end event?

> Maybe that doesn't have to hold this up though.

That was my thought.  Adding a lot of complexity for a feature we have no use case for.

garlick commented on src/modules/job-info/guest_watch.c:
```diff
@@ -607,32 +610,25 @@ static void guest_namespace_watch_continuation (flux_future_t *f, void *arg)
+            /* Guest namespace has been removed and eventlog has been
+             * moved to primary KVS namespace.  Fallthrough to primary
+             * KVS namespace to ensure that we have output all data to
+             * the user.  This is very small race possibility where
```
Is the "very small race possibility" documenting a known race or describing code that fixes the race?   It's right on top of another comment block describing what looks like a fix for a different race.

chu11 replied:
let me see how fixing up those comments.  There are really two races, and both are fixed in the same place.

garlick commented:
> could job state transitions be put into the eventlog? or perhaps another eventlog with a known end event?

You mean so we aren't replicating state machine details across modules?  At one point I had extracted the state machine code in libjob but I think that was part of a PR that didn't go in...

garlick commented:
I never got back to #1665, which was supposed to be a single job "state watch" service offered by the job manager.  It seemed like between the eventlog and the event state transition events, we had it covered.  Now  I wonder if this would be helpful to reduce complexity in job-info?

(sorry, definitely off topic for this PR.  Maybe just something to talk through next week)

chu11 commented:
Re-pushed with fixes based on comments.

- Added a new commit in the series that cleans up helper functions in `t2204-job-info` and `t2205-job-info-security`.

- I nixed issue #2356, since we can assume no eventlog can be empty.  The previous commit I had that fixed that issue is now gone.

- I created issue #2386 to describe the race between kvs-watch and job-info.

- The fix for #2338 (send ENODATA when eventlog at end) is spread out over multiple commits in this PR.  The fix for issue #2386 fixes a specific issue that happens to (almost entirely) overwrite one of the commits for #2338.  Since I fixed #2338 first and #2386 second, it didn't seem right to squash those commits together, so I left them separate.  That keeps the progression of fixes in the commit log.

- I modified the comments so they are hopefully more clear about racy bits in the code.
